File: 1002-maximum-width-ramp/1002-maximum-width-ramp.java
class Solution {
  public int maxWidthRamp(int[] nums) {
    int n = nums.length;
    int[] maxVal = new int[n];

    maxVal[n - 1] = nums[n - 1];

    for(int i = n-2;i>= 0;i--){
      maxVal[i] = Math.max(nums[i],maxVal[i+1]);
    }

    int l = 0, r = 0,mw = 0;

    while(l < n && r < n){
      while((l <= r) && (r < n) && nums[l] <= maxVal[r]){
        mw = Math.max(mw,(r-l));
        r++;
      }   
      l++;
    }
    return mw;
  }
}

File: 1002-maximum-width-ramp/1002-maximum-width-ramp.go
func maxWidthRamp(nums []int) int {
	n := len(nums)
	maxVal := make([]int, len(nums))

	maxVal[n-1] = nums[n-1]

	for i := len(nums) - 2; i >= 0; i-- {
		maxVal[i] = max(nums[i], maxVal[i+1])
	}
  

	l, r, mw := 0, 0, 0

	for r < n {
		for (l <= r) && (r < n) && (nums[l] <= maxVal[r]) {
      mw = max(mw,r-l)
      r++
		}
    l++
	}

  return mw
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}